File: serve.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# 本地启动：提供网页 + 转发 API（Ruby 版，无需 Python）
# 用法：ruby serve.rb

require "json"
require "net/http"
require "uri"
require "webrick"
require "thread"

PORT = Integer(ENV.fetch("PORT", "8080"))
ROOT = File.expand_path(__dir__)

MIME_TYPES = {
  ".html" => "text/html; charset=utf-8",
  ".js" => "application/javascript; charset=utf-8",
  ".css" => "text/css; charset=utf-8",
  ".json" => "application/json; charset=utf-8",
  ".png" => "image/png",
  ".jpg" => "image/jpeg",
  ".svg" => "image/svg+xml",
  ".ico" => "image/x-icon"
}.freeze

def mime_for(path)
  MIME_TYPES[File.extname(path).downcase] || "application/octet-stream"
end

class ChatServlet < WEBrick::HTTPServlet::AbstractServlet
  def do_OPTIONS(_req, res)
    res.status = 204
    set_cors(res)
  end

  def do_GET(req, res)
    path = req.path
    path = "/index.html" if path == "/" || path.empty?

    file = File.join(ROOT, path.sub(%r{\A/}, ""))
    unless file.start_with?(ROOT + File::SEPARATOR) && File.file?(file)
      res.status = 404
      res["Content-Type"] = "text/plain"
      res.body = "Not Found"
      set_cors(res)
      return
    end

    res.status = 200
    res["Content-Type"] = mime_for(file)
    res.body = File.binread(file)
    set_cors(res)
  end

  def do_POST(req, res)
    path = req.path.split("?", 2).first
    if path.end_with?("/chat/completions") || path.start_with?("/api/chat/completions")
      proxy_to(req, res, "chat/completions")
    elsif path.end_with?("/images/generations") || path.start_with?("/api/images/generations")
      proxy_to(req, res, "images/generations")
    elsif path.end_with?("/images/edit") || path.start_with?("/api/images/edit")
      proxy_to(req, res, "images/edit")
    else
      res.status = 404
      res["Content-Type"] = "text/plain"
      res.body = "Not Found"
      set_cors(res)
    end
  end

  private

  def set_cors(res)
    res["Access-Control-Allow-Origin"] = "*"
    res["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS"
    res["Access-Control-Allow-Headers"] = "Content-Type, Authorization, X-Target-Base-Url"
  end

  def json_error(res, code, message)
    payload = { error: { message: message } }.to_json
    res.status = code
    res["Content-Type"] = "application/json; charset=utf-8"
    res.body = payload
    set_cors(res)
  end

  def proxy_to(req, res, upstream_path)
    target_base = (req["X-Target-Base-Url"] || "").strip.sub(%r{/+\z}, "")
    if target_base.empty?
      json_error(res, 400, "缺少请求头 X-Target-Base-Url，请在设置中填写 Base URL")
      return
    end

    upstream = URI.join("#{target_base}/", upstream_path)
    http = Net::HTTP.new(upstream.host, upstream.port)
    http.use_ssl = upstream.scheme == "https"
    http.read_timeout = 600
    http.open_timeout = 30

    proxy_req = Net::HTTP::Post.new(upstream.request_uri)
    proxy_req["Content-Type"] = req["Content-Type"] if req["Content-Type"]
    proxy_req["Authorization"] = req["Authorization"] if req["Authorization"]
    proxy_req.body = req.body

    begin
      upstream_res = http.request(proxy_req)
      res.status = upstream_res.code.to_i
      upstream_res.each_header do |key, val|
        next if %w[connection transfer-encoding content-encoding].include?(key.downcase)

        res[key] = val
      end
      res.body = upstream_res.body
      set_cors(res)
    rescue StandardError => e
      json_error(res, 502, "无法连接中转站：#{e.message}")
    end
  end
end

server = WEBrick::HTTPServer.new(
  BindAddress: "127.0.0.1",
  Port: PORT,
  DocumentRoot: ROOT,
  Logger: WEBrick::Log.new($stderr, WEBrick::Log::INFO),
  AccessLog: [[$stderr, WEBrick::AccessLog::COMBINED_LOG_FORMAT]]
)
server.mount("/", ChatServlet)

url = "http://127.0.0.1:#{PORT}/"
puts "Chat 已启动: #{url}"
puts "按 Ctrl+C 停止"

Thread.new do
  sleep 0.8
  system("open", url)
end

trap("INT") { puts "\n已停止"; server.shutdown }

server.start

File: serve.py
#!/usr/bin/env python3
"""
本地启动：提供网页 + 转发 API，解决双击 HTML（file://）导致的 Failed to fetch / CORS。

用法：
  python3 serve.py
  或双击 Mac 上的「启动.command」
"""
from __future__ import annotations

import json
import os
import sys
import threading
import webbrowser
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

PORT = int(os.environ.get("PORT", "8080"))
ROOT = os.path.dirname(os.path.abspath(__file__))


class ChatHandler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def log_message(self, fmt, *args):
        sys.stderr.write("%s - - [%s] %s\n" % (self.address_string(), self.log_date_time_string(), fmt % args))

    def end_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header(
            "Access-Control-Allow-Headers",
            "Content-Type, Authorization, X-Target-Base-Url",
        )
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(204)
        self.end_headers()

    def do_GET(self):
        if self.path in ("/", ""):
            self.path = "/index.html"
        return super().do_GET()

    def do_POST(self):
        path = self.path.split("?", 1)[0]
        if path.rstrip("/").endswith("/chat/completions") or path.startswith("/api/chat/completions"):
            self._proxy_path("/chat/completions")
            return
        if path.rstrip("/").endswith("/images/generations") or path.startswith("/api/images/generations"):
            self._proxy_path("/images/generations")
            return
        self.send_error(404, "Not Found")

    def _proxy_path(self, suffix: str):
        target_base = (self.headers.get("X-Target-Base-Url") or "").strip().rstrip("/")
        if not target_base:
            self._json_error(400, "缺少请求头 X-Target-Base-Url，请在设置中填写 Base URL")
            return

        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length) if length else b""
        upstream = f"{target_base}{suffix}"

        headers = {"Content-Type": "application/json"}
        auth = self.headers.get("Authorization")
        if auth:
            headers["Authorization"] = auth

        req = Request(upstream, data=body, method="POST", headers=headers)

        try:
            with urlopen(req, timeout=600) as resp:
                self.send_response(resp.status)
                skip = {"connection", "transfer-encoding", "content-encoding"}
                for key, val in resp.headers.items():
                    if key.lower() not in skip:
                        self.send_header(key, val)
                self.end_headers()
                while True:
                    chunk = resp.read(8192)
                    if not chunk:
                        break
                    self.wfile.write(chunk)
        except HTTPError as e:
            err_body = e.read()
            self.send_response(e.code)
            ct = e.headers.get("Content-Type", "application/json")
            self.send_header("Content-Type", ct)
            self.end_headers()
            self.wfile.write(err_body)
        except URLError as e:
            self._json_error(502, f"无法连接中转站：{e.reason}")
        except Exception as e:
            self._json_error(500, str(e))

    def _json_error(self, code: int, message: str):
        payload = json.dumps({"error": {"message": message}}, ensure_ascii=False).encode()
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def main():
    os.chdir(ROOT)
    server = ThreadingHTTPServer(("127.0.0.1", PORT), ChatHandler)
    url = f"http://127.0.0.1:{PORT}/"
    print(f"Chat 已启动: {url}")
    print("按 Ctrl+C 停止")
    threading.Timer(0.8, lambda: webbrowser.open(url)).start()
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n已停止")
        server.server_close()


if __name__ == "__main__":
    main()
